includes/query/SMW_QueryProcessor.php

```diff
@@ -52,26 +53,31 @@ public static function setRecursiveTextProcessor( RecursiveTextProcessor $recurs
 	 * param name (string) => param value (mixed)
 	 *
 	 * @since 1.6.2
-	 * The return value changed in SMW 1.8 from an array with result values
-	 * to an array with Param objects.
+	 * The return value changed in SMW 3.1 from Param[] to ProcessedParam[]
 	 *
 	 * @param array $params
-	 * @param array $printRequests
+	 * @param PrintRequest[] $printRequests
 	 * @param boolean $unknownInvalid
 	 *
-	 * @return Param[]
+	 * @return ProcessedParam[]
 	 */
 	public static function getProcessedParams( array $params, array $printRequests = [], $unknownInvalid = true, $context = null, $showMode = false ) {
-		$validator = self::getValidatorForParams( $params, $printRequests, $unknownInvalid, $context, $showMode );
-		$validator->processParameters();
-		$parameters =  $validator->getParameters();
+		$validator = self::getProcessorForParams( $params, $printRequests, $unknownInvalid, $context, $showMode );
+		$processingResult = $validator->processParameters();
+
+		$parameters = [];
 
+		foreach ( $processingResult->getParameters() as $parameter ) {
+			$parameters[$parameter->getName()] = $parameter;
+		}
+
+		// TODO: ????
 		// Negates some weird behaviour of ParamDefinition::setDefault used in
 		// an individual printer.
 		// Applying $smwgQMaxLimit, $smwgQMaxInlineLimit will happen at a later
 		// stage.
 		if ( isset( $params['limit'] ) && isset( $parameters['limit'] ) ) {
-			$parameters['limit']->setValue( (int)$params['limit'] );
+//			$parameters['limit']->setValue( (int)$params['limit'] );
```

Comment:
The reason is outlined in #2474, if you think the cited issue no longer appears with the new `ParamDefinition` I'm fine with removing this line. Please confirm that the cited use case (as in #2474) continues to work otherwise this will introduce a regression.

Reply:
Thanks for linking. Now I get it.

There is no unexpected behavior on the ParamProcessor side. If validation for a parameter fails, it will be set to the default value. So if you set lower and/or upper bounds for a numeric parameter, and the value is out of bounds, it will go to the default and not to the nearest valid value.

If that is the behavior you want then simply do not set bounds in the parameter definition and handle this validation yourself (outside of ParamProcessor). If for whatever reason that is not possible, you can do something similar to what you did here.

Can't check now: why are bounds defined in the parameter definition? Are they always the same or does it depend on result printer?